### ISSUE TITLE
Peer review / Vertaisarviointi 2

### ISSUE THREAD

**Issue by sallasal** [ilkkamaksy/data-compression-algos#2]
# Peer review 2

The program was fetched from Github 26.4.2021, 8 PM.

The project starts smoothly. There were sample input in Main, and I replaced it with different inputs. All were encoded and decoded well with Huffman. But when this testing part is not necessary any more, it could be removed from the code?

Unfortunately, I was not able to compress the test file I created. The UI works well and I was able to choose my .txt test file (with content "TOBEORNOTTOBETOBEORNOTTOBE", I tried with and without quotation marks). But when I chose "Compress", I got this error message:

Exception in thread "JavaFX Application Thread" java.lang.StringIndexOutOfBoundsException: begin 0, end -1, length 8
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3319)
	at java.base/java.lang.String.substring(String.java:1874)
	at CompressionAlgorithms.utils.FileUtils.setInitialFileExtension(FileUtils.java:20)
	at CompressionAlgorithms.ui.MainScene.showSaveFileChooser(MainScene.java:129)
	at CompressionAlgorithms.ui.MainScene$SaveButtonListener.handle(MainScene.java:97)
	at CompressionAlgorithms.ui.MainScene$SaveButtonListener.handle(MainScene.java:93)
	...

Perhaps my file was not correctly created for the program. But I tried with different namings and contents, and that was always the result. So perhaps this could be checked out?

As I was not able to test the algorithms via GUI (could be my mistake), I checked also Lzw via Main. This is, I wrote short testing lines there. It seems that both algorithms, Huffman and Lzw, work well and return good values. Algorithms can also decompress the files correctly. So the core functionality seem to be working well! You had some problems with algorithms according to your documentation, but now it seems all is doing pretty good. Well solved.

As you have already implemented HashMap and Map yourself, you could also remove the imports from Lzw. AppService contains some unused imports too. Removing them would make the code even easier to read, and it would make clear that you use your own implementations.

Code is very well written and easy to understand. It was easy to understand what was done, and project structure was good. Testing and modifying the code was made easy for others.

## TL;DR

Algorithms seem to work well, and this is the core part. Some problems still with the file reader, if I have it right. Still, overall, very well done!

This is an interesting project, and I think it would be very nice to see also the testing results: which one is faster? Which one compresses files better?
